src/circuit.rs

```diff
@@ -120,6 +121,49 @@ pub struct Circuit {
     pub(crate) rcv: Value<ValueCommitTrapdoor>,
 }
 
+impl Circuit {
+    /// Constructs a `Circuit` from the following components:
+    /// - `merkle_path`: a merkle path witness of the `input_note`
+    /// - `input_note`: a note spent in scope of the action
+    /// - `output_note`: a note created in scope of the action
+    /// - `fvk`: full viewing key owning the `input_note`
+    /// - `alpha`: a scalar used for randomization of the action spend validating key
+    /// - `rcv`: trapdoor for the action value commitment
+    ///
+    /// Returns `None` if the `input_note` is not owned by the `fvk`.
+    pub fn from_action_context(
+        merkle_path: tree::MerklePath,
+        input_note: Note,
+        output_note: Note,
+        fvk: FullViewingKey,
+        alpha: pallas::Scalar,
+        rcv: ValueCommitTrapdoor,
+    ) -> Option<Circuit> {
+        let scope = fvk.scope_for_address(&input_note.recipient())?;
+        Some(Circuit {
+            path: Value::known(merkle_path.auth_path()),
+            pos: Value::known(merkle_path.position()),
+            g_d_old: Value::known(input_note.recipient().g_d()),
+            pk_d_old: Value::known(*input_note.recipient().pk_d()),
+            v_old: Value::known(input_note.value()),
+            rho_old: Value::known(input_note.rho()),
+            psi_old: Value::known(input_note.rseed().psi(&input_note.rho())),
+            rcm_old: Value::known(input_note.rseed().rcm(&input_note.rho())),
+            cm_old: Value::known(input_note.commitment()),
+            alpha: Value::known(alpha),
+            ak: Value::known(fvk.clone().into()),
+            nk: Value::known(*fvk.nk()),
+            rivk: Value::known(fvk.rivk(scope)),
+            g_d_new: Value::known(output_note.recipient().g_d()),
+            pk_d_new: Value::known(*output_note.recipient().pk_d()),
+            v_new: Value::known(output_note.value()),
+            psi_new: Value::known(output_note.rseed().psi(&output_note.rho())),
+            rcm_new: Value::known(output_note.rseed().rcm(&output_note.rho())),
+            rcv: Value::known(rcv),
+        })
+    }
+}
```

Comment:
```suggestion
    pub fn from_action_context(
        spend: SpendInfo,
        output_note: Note,
        alpha: pallas::Scalar,
        rcv: ValueCommitTrapdoor,
    ) -> Option<Circuit> {
        (spend.note.nullifier(&spend.fvk) == output_note.rho()).then(|| {
            Self::from_action_context_unchecked(
                spend_info,
                output_note,
                alpha,
                rcv,
            )
        })
    }

    pub(crate) fn from_action_context_unchecked(
        spend: SpendInfo,
        output_note: Note,
        alpha: pallas::Scalar,
        rcv: ValueCommitTrapdoor,
    ) -> Circuit {
        let scope = fvk.scope_for_address(&input_note.recipient())?;

        let sender_address = spend.note.recipient();
        let rho_old = spend.note.rho();
        let psi_old = spend.note.rseed().psi(&rho_old);
        let rcm_old = spend.note.rseed().rcm(&rho_old);

        let rho_new = output_note.rho();
        let psi_new = output_note.rseed().psi(&rho_new);
        let rcm_new = output_note.rseed().rcm(&rho_new);

        Circuit {
            path: Value::known(spend.merkle_path.auth_path()),
            pos: Value::known(spend.merkle_path.position()),
            g_d_old: Value::known(sender_address.g_d()),
            pk_d_old: Value::known(*sender_address.pk_d()),
            v_old: Value::known(input_note.value()),
            rho_old: Value::known(rho_old),
            psi_old: Value::known(psi_old),
            rcm_old: Value::known(rcm_old),
            cm_old: Value::known(spend.note.commitment()),
            alpha: Value::known(alpha),
            ak: Value::known(fvk.clone().into()),
            nk: Value::known(*fvk.nk()),
            rivk: Value::known(fvk.rivk(scope)),
            g_d_new: Value::known(output_note.recipient().g_d()),
            pk_d_new: Value::known(*output_note.recipient().pk_d()),
            v_new: Value::known(output_note.value()),
            psi_new: Value::known(psi_new),
            rcm_new: Value::known(rcm_new),
            rcv: Value::known(rcv),
        })
    }
}
```

Reply:
Mind you that `fvk.scope_for_address(&input_note.recipient())?;` makes `from_action_context_unchecked` to failable.

Reply:
It wasn't redundant because `scope` is used in `fvk.rivk(scope)`. You are correct that `from_action_context_unchecked` needs to be fallible, as long as it still calls `scope_for_address`.

I think I fixed it. However I don't really like that the two possible errors are collapsed. @str4d, can you think of a better approach, or should we change it to return a `Result`?

Reply:
I fixed the suggestion just to convince myself, but you'd already come up with the same thing.